### ISSUE TITLE
深度学习基础知识

### ISSUE THREAD

**Issue by zhangjun** [zhangjun/zhangjun.github.io#31]
# 归一化方法

![](https://img-blog.csdnimg.cn/20210319221042880.png)

## Batch Norm
Batch Norm在通道维度进行归一化，最后得到C个统计量u，δ。假设输入特征为[N, H, W, C]，在C的每个维度上对[N, H, W]计算其均值、方差，用于该维度上的归一化操作。
```python
import numpy as np
import torch
import torch.nn as nn
from einops import rearrange, repeat, reduce

image = [np.random.randn(30, 40, 3) for _ in range(16)]
image = rearrange(image, 'b h w c -> b h w c')
# print(rearrange(image, 'b h w c -> b h w c').shape)

image_ = rearrange(image, 'b h w c -> (b h w) c')
mean = rearrange(image_.mean(axis=0), 'c -> 1 1 1 c')
std = rearrange(image_.std(axis=0), 'c -> 1 1 1 c')

y_ =  (image - mean)/std

b, h, w, c = image.shape
bn = nn.BatchNorm2d(c, eps=1e-10, affine=False, track_running_stats=False)
y = bn(torch.from_numpy(image))

print('diff={}\n'.format(torch.abs(y - y_).max()))

```
## Layer Norm
Layer Norm以样本为单位计算统计量，因此最后会得到N个u，δ。假设输入特征为[N, H, W, C]，在N的每个维度上对[H, W，C]计算其均值、方差，用于该维度上的归一化操作。
```python
import numpy as np
import torch
import torch.nn as nn
from einops import rearrange, repeat, reduce

x = torch.randn((6, 3, 20, 20))
b, c, h, w = x.shape

layer_norm = nn.LayerNorm([c, h, w], eps=1e-12, elementwise_affine=False)
y = layer_norm(x)

x_ = rearrange(x, 'b c h w -> (h w c) b')
mean = rearrange(x_.mean(axis=0), 'b -> b 1 1 1')
std = rearrange(x_.std(axis=0), 'b -> b 1 1 1')

y_ =  (x - mean)/std

print('diff={}\n'.format(torch.abs(y - y_).max()))

```
## Instance Norm
```python
import numpy as np
import torch
import torch.nn as nn
from einops import rearrange, repeat, reduce

x = torch.randn((6, 3, 20, 20))
b, c, h, w = x.shape

instance_norm = nn.InstanceNorm2d(c, eps=1e-12, affine=False, track_running_stats=False)
y = instance_norm(x)

x_ = rearrange(x, 'b c h w -> b c (h w)')
# mean = rearrange(x_.mean(axis=2), 'b c -> b c 1 1')
# std = rearrange(x_.std(axis=2), 'b c -> b c 1 1')
mean = rearrange(x_.mean(dim=2), 'b c -> b c 1 1')
std = rearrange(x_.std(dim=2), 'b c -> b c 1 1')

y_ =  (x - mean)/std

print('diff={}\n'.format(torch.abs(y - y_).max()))

```
## Group Norm
```python
import numpy as np
import torch
import torch.nn as nn
from einops import rearrange, repeat, reduce

x = torch.randn((6, 6, 20, 20))
b, c, h, w = x.shape
group_num = 3
n = 2

group_norm = nn.GroupNorm(group_num, c, eps=1e-12, affine=False)
y = group_norm(x)

x_ = rearrange(x, 'b (g n) h w -> b g (n h w)', g = group_num) # [6, 3, 2*20*20]
print(x_.shape)
mean = rearrange(x_.mean(dim=2), 'b g -> b g 1')  # [6, 3, 1]
std = rearrange(x_.std(dim=2), 'b g -> b g 1')

y_ =  (x_ - mean)/std
y_ = rearrange(y_, 'b g (n h w) -> b (g n) h w', g = group_num, h = h, w = w)

print('diff={}\n'.format(torch.abs(y - y_).max()))

```

**Comment by zhangjun** [comment]
# parallelism
https://colossalai.org/docs/concepts/paradigms_of_parallelism/
https://www.zhihu.com/question/508671222/answer/2290801813
https://www.cnblogs.com/marsggbo/p/16871789.html
## Data Parallel
all-reduce
<img src="https://github.com/zhangjun/zhangjun.github.io/assets/1312389/09dead4f-0a40-4347-a3c5-978fdc0d4051" width="250"/>
## Model Parallel
### Tensor Parallel
all-gather
### Pipeline Parallel
model is split by layer into several chunks, each chunk is given to a device.
![image](https://github.com/zhangjun/zhangjun.github.io/assets/1312389/dc9f79c9-bd39-4b2f-8c01-30cd3d0c844d)
<img src="https://github.com/zhangjun/zhangjun.github.io/assets/1312389/b52e24e0-9f6d-4800-8ae5-8852672908ec" width="250"/>
## Optimizer-Level Parallel
## 重计算(Recomputation or Checkpointing)
## 零冗余优化器 (Zero REdundancy Optimizer)
## 专家并行 MOE
<img src="https://github.com/zhangjun/zhangjun.github.io/assets/1312389/dda0a6f1-5594-4f66-93be-86e1374e893c" width="400"/>

**Comment by zhangjun** [comment]
# MultiHeadAttention融合实现
- 实现一: 训练算子、大模型FP16推理使用
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/fused/fused_attention_op.cu
- 实现二、推理融合算子
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/fused/multihead_matmul_op.cu
- 实现三、TensorRT推理
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/tensorrt/convert/multihead_matmul_op.cc
